### DISCUSSION TITLE
Basic precedence question

### DISCUSSION THREAD

**Discussion by psnszsn** [tree-sitter/tree-sitter#1440] [Q&A] [answered]
I'm writing a parser for a simple language which contains multiple variable assignments. The rule for a variable and its value looks something like this:
```js
        variable_assignment: $ => seq(
            field('name',$.word),
            field('operator', '='),
            optional(field('value', $.value)),
            \/n\,
        ),
```

However, for a few given variable names, I want to capture them separately.

```js
        PATH_assignment: $ => seq(
            'PATH',
            field('operator', '='),
            optional(field('value', $.value)),
            \/n\,
        ),
```

How can I give precedence to the PATH assignment instead of the more general one?

I tried this as well as a bunch of other variations but nothing seems to work:
```js
        assignment: $ => choice(
            prec(2, $.PATH_assignment),
            prec(1, $.variable_assignment),
        ),
```

**Comment by ahelwer** [comment, marked as answer]
It should work without precedence annotations since `'PATH'` is a more specific match than `$.word` according to the ever-useful [conflicting tokens](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens) rules. Can you give an example input that fails for this small test grammar?
Also why is the value optional?

**Reply by psnszsn** [reply]
It seems like I oversimplified a bit too much. This is a grammar.js file that is closer to what I'm trying to accomplish:

```js
module.exports = grammar({
  name: 'test',
  word: $ => $.word,
  extras: $ => [/\s/, alias(token(seq('\\', /\r?\n|\r/)), '\\')],

  rules: {
    root: $ => repeat($._statement),

    _statement: $ =>
      choice(
        prec(-1, $.variable_assignment),
        prec(1, $.special_variable_assignment),
      ),

    variable_assignment: $ =>
      seq(
        field("var_name", /[\w_\-\/\.]+/),
        field("operator", '='),
        optional(field("var_value", $.value)),
        /\n/
      ),

    special_variable_assignment: $ =>
      seq(
        field("var_name", /[\w_\-\/]+/),
        token.immediate('.'),
        token.immediate(choice('PATH', 'ETC')),
        field("operator", '='),
        optional(field("var_value", $.value)),
        /\n/
      ),

    value: $ => /[\w\-\.\/_=+${},()]+/,
    word: $ => /[\w\-\/_]+/,

  },
})

```
The value is optional since omitting it is a way to unset the variable.

I'm using this as input:

```
TEST.PATH =
TEST.ETC = value
```
I get this output:
```
(root [0, 0] - [3, 0]
  (variable_assignment [0, 0] - [1, 0])
  (variable_assignment [1, 0] - [3, 0]
    var_value: (value [1, 11] - [1, 16])))
```

Desired output:
```
(root [0, 0] - [3, 0]
  (special_variable_assignment [0, 0] - [1, 0])
  (special_variable_assignment [1, 0] - [3, 0]
    var_value: (value [1, 11] - [1, 16])))
```

**Reply by ahelwer** [reply]
Okay. So what's happening is your `var_name` regex in `variable_assignment` `/[\w_\-\/\.]+/` is gobbling up all of `TEST.PATH`. This is a lexical precedence issue, not a parser precedence issue - so your precedence annotations in `_statement` aren't doing anything. In order to give lexical precedence instructions to the lexer instead of the parser you have to put them inside the `token` function. So if you rewrite `special_variable_assignment` as follows it seems to do what you want:
```js
special_variable_assignment: $ =>
  seq(
    field("var_name", token(
      prec(1, seq(
        /[\w_\-\/]+/,
        '.',
        choice('PATH', 'ETC')
      )
    ))),
    field("operator", '='),
    optional(field("var_value", $.value)),
    /\n/
  ),
```
you can think of the `token` function as enabling you to use `seq` and `choice` as a sort of DSL alternative to regexes - don't need to mess around with `token.immediate` or anything. As documented in rule 2 of the [conflicting tokens documentation](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens), using the `prec` function inside the `token` function specifices lexical precedence instead of parser-level precedence.

**Reply by psnszsn** [reply]
Thanks! That did the trick!

**Reply by md143652** [reply]
That cleared up a lot for me, too. Thank you!

**Reply by psnszsn** [reply]
I bumped into another issue related to this. I want to assign field names to the children inside `token()`, but it doesn't work since everything inside the `token` function is excluded from the syntax tree. How can this be solved?

**Reply by ahelwer** [reply]
@psnszsn if you want to break apart a token into multiple named tokens, then it isn't a token - it's a series of tokens. You have to use parser-level precedence annotations.

**Reply by psnszsn** [reply]
@ahelwer then I'm back to the example from [my first comment](https://github.com/tree-sitter/tree-sitter/discussions/1440#discussioncomment-1472241).  Any idea what I did wrong there?

**Reply by ahelwer** [reply]
Split the other regex up into a series of tokens too. Ensure the first token in the series is the same for both. You can hide the sub-tokens by prepending `_` to their rule names if you want, or doing the same with `alias`.
